contracts/stake-cw20/src/contract.rs

```diff
@@ -66,42 +75,52 @@ pub fn execute(
         ExecuteMsg::Receive(msg) => execute_receive(deps, env, info, msg),
         ExecuteMsg::Unstake { amount } => execute_unstake(deps, env, info, amount),
         ExecuteMsg::Claim {} => execute_claim(deps, env, info),
-        ExecuteMsg::UpdateConfig { admin, duration } => {
-            execute_update_config(info, deps, admin, duration)
-        }
+        ExecuteMsg::UpdateConfig {
+            owner,
+            manager,
+            duration,
+        } => execute_update_config(info, deps, owner, manager, duration),
+        ExecuteMsg::AddHook { addr } => execute_add_hook(deps, env, info, addr),
+        ExecuteMsg::RemoveHook { addr } => execute_remove_hook(deps, env, info, addr),
     }
 }
 
 pub fn execute_update_config(
     info: MessageInfo,
     deps: DepsMut,
-    new_admin: Option<Addr>,
+    new_owner: Option<Addr>,
+    new_manager: Option<Addr>,
     duration: Option<Duration>,
 ) -> Result<Response, ContractError> {
     let mut config: Config = CONFIG.load(deps.storage)?;
-    match config.admin {
-        None => Err(ContractError::NoAdminConfigured {}),
-        Some(current_admin) => {
-            if info.sender != current_admin {
-                return Err(ContractError::Unauthorized {
-                    expected: current_admin,
-                    received: info.sender,
-                });
-            }
+    if Some(info.sender.clone()) != config.owner && Some(info.sender.clone()) != config.manager {
+        return Err(ContractError::Unauthorized {});
+    };
+    if Some(info.sender) != config.owner && new_owner != config.owner {
+        return Err(ContractError::OnlyOwnerCanChangeOwner {});
+    };
 
-            config.admin = new_admin;
-            config.unstaking_duration = duration;
-
-            CONFIG.save(deps.storage, &config)?;
-            Ok(Response::new().add_attribute(
-                "admin",
-                config
-                    .admin
-                    .map(|a| a.to_string())
-                    .unwrap_or_else(|| "None".to_string()),
-            ))
-        }
-    }
+    config.owner = new_owner;
+    config.manager = new_manager;
```

Comment:
Probably want to validate these addresses.

Reply:
Nope :/ https://docs.rs/cosmwasm-std/latest/cosmwasm_std/struct.Addr.html

Reply:
validated